### PR TITLE
Adding text memories offline

### PR DESCRIPTION
**Pull request by pbywater**
- Changes deletePendingMemories helper to processPendingMemories to deal with both adding and deleting
- Saves memory to local storage and generates a temporary fake id to allow memory to appear in the DOM
- Makes the POST request when back online and removes temporary memories from local storage

Relates to #163

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: src/client/d3/d3.js [hunk 1]
```diff
@@ -217,16 +217,30 @@ function render(updatedData) {
...
+          }, animationDuration);
+        },
+      });
+    } else {
+      const data = $('#add-text-form').serialize();
+      const splitData = data.split('&');
+      const heading = splitData[0].split('=')[1];
+      const text = splitData[1].split('=')[1];
+      const tag = splitData[2].split('=')[1];
+      const id = getRandomInt(10000, 999999);
```
Comment (des-des) [inline src/client/d3/d3.js:237]: Not convinced by this approach to id generation

File: src/client/d3/d3.js [hunk 2]
```diff
@@ -217,16 +217,30 @@ function render(updatedData) {
...
+        success: () => {
+          setTimeout(() => {
+            update();
+          }, animationDuration);
+        },
+      });
+    } else {
+      const data = $('#add-text-form').serialize();
+      const splitData = data.split('&');
+      const heading = splitData[0].split('=')[1];
```
Comment (des-des) [inline src/client/d3/d3.js:234]: Dont do this, just get the data from each field separately.

Comment (des-des) [reply src/client/d3/d3.js:234]: ie heading = $('sel').value

File: src/client/d3/d3.js (hunk 1 above)
Comment (des-des) [reply src/client/d3/d3.js:237]: 1. Use uuid or something similar for client side id generation.
2. Either generate ids on the server, or on the client, do not do some mixture of the 2.


If `2` is too difficult, you could try and generate some kind of temporary id on the client, then replace it with the one generated by the db..

File: src/client/helpers/helpers.js [hunk 3]
```diff
@@ -287,6 +296,21 @@ function deletePendingMemories(cb) {
     cb();
   })
 }
+if(localStorage.getItem('textToAdd') !== null) {
+const newTextMemories = JSON.parse(localStorage.getItem('textToAdd'));
```
Comment (des-des) [inline src/client/helpers/helpers.js]: indentation

File: src/client/helpers/helpers.js [hunk 4]
```diff
@@ -287,6 +296,21 @@ function deletePendingMemories(cb) {
     cb();
```
Comment (des-des) [inline src/client/helpers/helpers.js:296]: ?? How many times does this callback get called

File: src/client/helpers/helpers.js [hunk 5]
```diff
@@ -287,6 +296,21 @@ function deletePendingMemories(cb) {
...
+  const text = memory.text;
+  const heading = memory.heading;
+  const tag = memory.tag;
+  $.ajax({
+    method: 'POST',
+    url: 'memory-input-text',
+    data: {memory_text: text, heading, tag},
+    success: () => clearPendingActions('textToAdd', index),
+  });
+  cb();
```
Comment (des-des) [inline src/client/helpers/helpers.js]: How many times does this callback get called / this does not make any sense / this is broken

File: src/client/helpers/helpers.js [hunk 6]
```diff
@@ -261,6 +261,15 @@ function removeMemoryFromStoredData(id) {
 return offlineData;
 }
 
+function addMemoryToStoredData(id, heading, text, tag) {
+  const offlineData = JSON.parse(localStorage.getItem('data'));
+    const toAdd = {heading, id, likes: 0, media_type: "text_only", memory_text: text, tag, visits:0, memory_asset_url:''};
+  offlineData.push(toAdd);
+  const offlineDataAfterAdding = JSON.stringify(offlineData);
+  localStorage.setItem('data', offlineDataAfterAdding);
+return offlineData;
```
Comment (des-des) [inline src/client/helpers/helpers.js]: indentation

**Review by des-des — APPROVED** [review]

File: src/client/d3/d3.js (hunk 1 above)
Comment (pbywater) [reply src/client/d3/d3.js:237]: This is just temporary - a real id is created in the normal way once the user goes back online

File: src/client/helpers/helpers.js (hunk 4 above)
Comment (pbywater) [reply src/client/helpers/helpers.js:296]: Every time something is deleted I think! Haha - I think we discussed before and said it wasn't ideal but would be ok due to time constraints?

File: src/client/d3/d3.js (hunk 2 above)
Comment (pbywater) [reply src/client/d3/d3.js:234]: Tried this but I'm getting undefined :(

File: src/client/helpers/helpers.js (hunk 5 above)
Comment (pbywater) [reply src/client/helpers/helpers.js]: 😭  - it's rerendering after each update so if there are several updates, there are several transitions. Is there a better way to do it?

File: src/client/d3/d3.js (hunk 1 above)
Comment (des-des) [reply src/client/d3/d3.js:237]: :+1:

**Review by des-des — APPROVED** [review]

**Pull request merged**